Verona/Verona.swift

```diff
@@ -102,17 +102,73 @@ public enum NSAttributedStringAttribute {
 
 public class NSAttributedStringAttributeBuilder {
     
-    private var attributesDictionary: [String : AnyObject] = [ : ]
+    private var clearAttributesOnNextString: Bool = false
+    
+    private var currentString: String = ""
+    private var strings = [String]()
+    
+    var currentAttributes: [String : AnyObject] = [ : ]
+    var attributes = [[String: AnyObject]]()
     
     public init() { }
     
-    public func setAttribute(attribute: NSAttributedStringAttribute) {
+    // Builder functions
+    
+    public func setAttribute(attribute: NSAttributedStringAttribute) throws {
         let keyValuePair = attribute.attributedStringKeyValuePair
-        attributesDictionary.updateValue(keyValuePair.value, forKey: keyValuePair.key)
+        currentAttributes.updateValue(keyValuePair.value, forKey: keyValuePair.key)
     }
     
-    public func attributedStringAttributesDictionary() -> [String: AnyObject]? {
-        return attributesDictionary
+    public func clearAttributes() {
+        currentAttributes.removeAll()
+    }
+    
+    public func addAttributes(attributes: [String: AnyObject]) {
+        for (key, value) in attributes {
+            currentAttributes.updateValue(value, forKey: key)
+        }
+    }
+    
+    public func removeAttribute(attribute: NSAttributedStringAttribute) throws {
+        let keyValuePair = attribute.attributedStringKeyValuePair
+        currentAttributes.removeValueForKey(keyValuePair.key)
+    }
+    
+    public func removeAttributes(attributes: [String: AnyObject]) {
+        for (key, _) in attributes {
+            currentAttributes.removeValueForKey(key)
+        }
+    }
+    
+    public func nextString(string: String) {
+        endCurrent()
+        currentString = string
+    }
+    
+    
+    // Internal helpers
+    
+    private func endCurrent() {
+        attributes.append(currentAttributes)
+        strings.append(currentString)
+        if (clearAttributesOnNextString) {
+            clearAttributes()
+        }
+        currentString = ""
+    }
+    
+    private func finalize() -> NSAttributedString {
+        let attrStrings = zip(strings, attributes)
+            .map { (str: String, attr: [String: AnyObject]) in
+                return NSAttributedString(string: str, attributes: attr)
+            }
+        
+        // TODO: Get reduce to work instead of whats below. Was showing ambiguous errors with mutable string as accumulator
```

Comment:
Not 100% sure if `reduce` will work

what about something like this to encapsulate the logic

``` Swift
        let finalString: NSAttributedString = {
            let mut = NSMutableAttributedString()
            attrStrings.forEach{ mut.appendAttributedString($0) }
            return NSAttributedString(attributedString: mut)
        }()
        
        return finalString
```